File: trialcode/fractal.js
timer=timer=require("ringo/utils/debug").timer;

// function radians(degrees){return java.lang.Math.toRadians(degrees)}
radians = java.lang.Math.toRadians;


function draw_tree(g2d,length,depth){
  if(depth > 0){
    g2d.drawLine(0, 0, length, 0);

    g2d.translate(length,0);

    g2d.rotate(radians(-30));
    draw_tree(g2d, 0.75*length, depth-1);
    g2d.rotate(radians(30));

    g2d.rotate(radians(30));
    draw_tree(g2d, 0.75*length, depth-1);
    g2d.rotate(radians(-30));

    g2d.translate(-length,0);

  }
}

function draw(g2d){
  var depth = 16;
  var length = 100;
  print("Depth: " + depth);
  timer(function(){draw_tree(g2d, length, depth)});
  // var start = java.lang.System.nanoTime();
  // draw_tree(g2d, length, depth);
  // var stop = java.lang.System.nanoTime();
  // print(Math.round((stop - start) / 1000000), "millis");

}

width=800;
height=800;
var gg2d;

frame = new javax.swing.JFrame();
pan = new JavaAdapter(javax.swing.JPanel,
  {
    paint:function(g2d){
      g2d.translate( width/2, height/2);
      g2d.rotate(radians(-90));
      draw(g2d);
      // draw = function(g2d){};
  }
})


frame.setContentPane(pan);
frame.pack();
frame.setSize(width,height);
frame.show();

// pan = new JavaAdapter(javax.swing.JFrame,
//   {
//     paint:function(g2d){
//       g2d.translate( width/2, height/2);
//       g2d.rotate(radians(-90));
//       draw(g2d);
//       // draw = function(g2d){};
//   }
// })


// frame.pack();
// frame.setSize(width,height);
// frame.show();

File: trialcode/fractal.clj
(import
 '(javax.swing JFrame JPanel)
 '(java.awt Graphics2D))

(def wof *warn-on-reflection* )
(set! *warn-on-reflection* true)

(defn radians [degrees] (. java.lang.Math toRadians degrees))

(defmacro rot [^Graphics2D g2d angle & body]
  `(do (. ~g2d rotate (radians ~angle))
       (let [a# (do ~@body)]
	 (. ~g2d rotate (radians (- 0 ~angle)))
	 a#)))
(defmacro trans [^Graphics2D g2d dx dy & body]
  `(do (. ~g2d translate ~dx ~dy)
       (let [a# (do ~@body)]
	 (. ~g2d translate  (int (- 0 ~dx)) (int (- 0 ~dy)))
	 a#)))
					; -- API END --


					; -- CLIENT CODE START --
(def width 400)
(def height 400)


(defn draw_tree [^Graphics2D g2d length depth]
  (if (> depth 0)
    (do
      (. g2d drawLine 0 0 length 0)
      (trans g2d (int length) 0
	     (rot g2d -30
		  (draw_tree g2d (* length 0.75) (- depth 1)))
	     (rot g2d 30
		  (draw_tree g2d (* length 0.75) (- depth 1)))))))


;; (defn draw_tree [^Graphics2D g2d length depth]
;;   (if (> depth 0)
;;     (let [z (int 0)
;; 	  len (int length)
;; 	  r1 (double (radians -30))
;; 	  r2 (double (radians 30))]

;;       (.drawLine g2d 0 0 length 0)
;;       (.translate g2d len z)

;;       (.rotate g2d r1)
;;       (draw_tree g2d (* length 0.75) (- depth 1))
;;       (.rotate g2d r2)

;;       (.rotate g2d r2)
;;       (draw_tree g2d (* length 0.75) (- depth 1))
;;       (.rotate g2d r1)

;;       (.translate g2d (- len) z))))

(defn draw [^Graphics2D g2d]
  (doto g2d
    (. translate (int (/ width 2)) (int (/ height 2)))
    (. rotate (radians -90)))
  (time (draw_tree g2d 50 13)))
; -- CLIENT CODE END --


; -- API START --  
(def frame (new JFrame))
(def display
     (proxy [JPanel] []
       (paintComponent [g2d]  (draw g2d))))

;; (.setDefaultCloseOperation frame
;;   WindowConstants/EXIT_ON_CLOSE)
(doto frame
  (.setContentPane display)
  (.setTitle "Clojure is Ornery") 
  (.pack )
  (.setSize width height)
  (.show)
  
  (println "\n\nStarting up" (.getTitle frame) "\n"))
					; -- API END --
(set! *warn-on-reflection* wof)
(ns-unmap 'user 'wof)